app/src/androidTest/java/org/simple/clinic/storage/DaoWithUpsertAndroidTest.kt

```diff
@@ -0,0 +1,66 @@
+package org.simple.clinic.storage
+
+import com.google.common.truth.Truth.assertThat
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.simple.clinic.AppDatabase
+import org.simple.clinic.TestClinicApp
+import org.simple.clinic.TestData
+import org.simple.clinic.patient.Gender.FEMALE
+import org.simple.clinic.patient.Gender.MALE
+import java.util.UUID
+import javax.inject.Inject
+
+class DaoWithUpsertAndroidTest {
+
+  @Inject
+  lateinit var database: AppDatabase
+
+  @Inject
+  lateinit var testData: TestData
+
+  @Before
+  fun setUp() {
+    TestClinicApp.appComponent().inject(this)
+  }
+
+  @After
+  fun tearDown() {
+    database.clearAllTables()
+  }
+
+  @Test
+  fun when_upserting_a_record_it_should_be_inserted_if_it_doesnt_exist_already() {
+    val patientUuid = UUID.randomUUID()
```

Comment:
@vinaysshenoy @sanchita what do you think of these tests? `DaoWithUpsert` is abstract so only a subclass can be tested.

Reply:
I know. The implementation is generated by Room so I had to use an actual implementation. Any suggestions?

Reply:
I'd say not call this a `DaoWithUpsertAndroidTest` since it indicates this is a test for generic implementations of this Dao. I'd suggest naming it `PatientDaoUpsertTest` instead. Or even moving these tests to the `PatientRepositoryAndroidTest` and getting rid of this class altogether.